### PR TITLE
Ensure a main class element is findable in Scala code passed by `stdin`

### LINKED ISSUE #907: Scala CLI does not see main classes from stdin passed as `_.scala`
**Version(s)**
`0.1.3`

**Describe the bug**

When provide a valid scala main class on stdin and run scala-cli it does not see any main classes.

**To Reproduce**

```
echo "@main def a = println(1)" | scala-cli run  _.scala
```

**Expected behaviour**
The main class should be found and the code should run

### PR DESCRIPTION
**Pull request by Gedochao**
Fixes #907

### CODE REVIEW

**Review by lwronski — COMMENTED** [review]
I think, we should fix following use case and update scala-cli integration test
```
lwronski@VL-D-0317 scala-cli % scala-cli version                         
0.1.4
lwronski@VL-D-0317 scala-cli % scala-cli https://gist.github.com/lwronski/ab522c25905a50d07d0e4cb3e1fcf5d5
Hello
lwronski@VL-D-0317 scala-cli % git branch
* fix-main-class-for-stdin-scala-code
  main
lwronski@VL-D-0317 scala-cli % ./mill scala https://gist.github.com/lwronski/ab522c25905a50d07d0e4cb3e1fcf5d5
Compiling project (Scala 3.1.1, JVM)
Warning: Flag -bootclasspath set repeatedly
Warning: Flag -classpath set repeatedly
[error] ab522c25905a50d07d0e4cb3e1fcf5d5-main/Hello.scala:3:11: Not found: Messages
Error compiling project (Scala 3.1.1, JVM)
Compilation failed
1 targets failed
cli.run subprocess failed
```

**Comment by Gedochao** [comment]
> I think, we should fix following use case and update scala-cli integration test
> 
> ```
> lwronski@VL-D-0317 scala-cli % scala-cli version                         
> 0.1.4
> lwronski@VL-D-0317 scala-cli % scala-cli https://gist.github.com/lwronski/ab522c25905a50d07d0e4cb3e1fcf5d5
> Hello
> lwronski@VL-D-0317 scala-cli % git branch
> * fix-main-class-for-stdin-scala-code
>   main
> lwronski@VL-D-0317 scala-cli % ./mill scala https://gist.github.com/lwronski/ab522c25905a50d07d0e4cb3e1fcf5d5
> Compiling project (Scala 3.1.1, JVM)
> Warning: Flag -bootclasspath set repeatedly
> Warning: Flag -classpath set repeatedly
> [error] ab522c25905a50d07d0e4cb3e1fcf5d5-main/Hello.scala:3:11: Not found: Messages
> Error compiling project (Scala 3.1.1, JVM)
> Compilation failed
> 1 targets failed
> cli.run subprocess failed
> ```

Should be fixed now. The issue involved wrongly passing the `relPath` appropriate for a `stdin` input to multiple files inside of a zip archive.
I added a relevant integration test.

File: modules/build/src/main/scala/scala/build/preprocessing/ScalaPreprocessor.scala [hunk 1]
```diff
@@ -60,7 +60,7 @@ case object ScalaPreprocessor extends Preprocessor {
     UsingPublishDirectiveHandler
   )
 
-  val requireDirectiveHandlers = Seq[RequireDirectiveHandler](
+  val requireDirectiveHandlers: Seq[RequireDirectiveHandler] = Seq[RequireDirectiveHandler](
```
Comment (alexarchambault) [inline modules/build/src/main/scala/scala/build/preprocessing/ScalaPreprocessor.scala]: Do we really need this? There's already a `Seq[RequireDirectiveHandler]` on the same line…

File: modules/build/src/main/scala/scala/build/preprocessing/ScalaPreprocessor.scala [hunk 2]
```diff
@@ -123,15 +128,15 @@ case object ScalaPreprocessor extends Preprocessor {
...
-            0,
-            Some(options),
-            Some(requirements),
+            ignoreLen = 0,
+            options = Some(options),
+            requirements = Some(requirements),
             scopedRequirements,
-            None,
-            v.scopePath
+            mainClassOpt = Some(className),
```
Comment (alexarchambault) [inline modules/build/src/main/scala/scala/build/preprocessing/ScalaPreprocessor.scala:138]: Why do we need this? As it's a Scala file (not a script), we shouldn't have to come up with a main class name here. If users defined one, it should be found after compilation, doesn't it? (at least, that's how it should work in theory)

Comment (Gedochao) [reply modules/build/src/main/scala/scala/build/preprocessing/ScalaPreprocessor.scala:138]: The short answer is, if we don't provide a fallback main class name here, a `NoMainClassFoundError` gets thrown in[ `scala.build.Build`
](https://github.com/VirtusLab/scala-cli/blob/main/modules/build/src/main/scala/scala/build/Build.scala#L63)
We create [an inferred class name for Inputs.ScalaFile](https://github.com/VirtusLab/scala-cli/blob/main/modules/build/src/main/scala/scala/build/preprocessing/ScalaPreprocessor.scala#L75) in a similar fashion, which my intension was to stay consistent with.
In turn, the `mainClassOpt`  seems to be used as fallback for the found main class in [`scala.build.Build`](https://github.com/VirtusLab/scala-cli/blob/main/modules/build/src/main/scala/scala/build/Build.scala#L60)

It seems to me that if we don't need to set the `mainClassOpt` here (and rely on the compilation instead), then we probably shouldn't need the inferred class name for `Inputs.ScalaFile`, either?
Is passing the class name here meant solely for the Scala script wrapper?

File: modules/build/src/main/scala/scala/build/preprocessing/ScalaPreprocessor.scala (hunk 1 above)
Comment (Gedochao) [reply modules/build/src/main/scala/scala/build/preprocessing/ScalaPreprocessor.scala]: right, no need to pass the type param twice, in the same line, my bad.
should be better now.

**Review by alexarchambault — APPROVED** [review]

**Review by lwronski — APPROVED** [review]

**Pull request merged**